pkg/router/controller/router_controller.go

```diff
@@ -227,6 +228,59 @@ func (c *RouterController) HandleEndpoints(eventType watch.EventType, obj interf
 	c.Commit()
 }
 
+// HandleEndpointSlice handles a single EndpointSlice event and refreshes the router backend.
+func (c *RouterController) HandleEndpointSlice(eventType watch.EventType, objMeta metav1.ObjectMeta, items []discoveryv1beta1.EndpointSlice) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	var subsets []kapi.EndpointSubset
+
+	for i := range items {
+		var ports []kapi.EndpointPort
+		var addresses []kapi.EndpointAddress
+
+		for j := range items[i].Endpoints {
+			for k := range items[i].Endpoints[j].Addresses {
+				epa := kapi.EndpointAddress{
+					IP:        items[i].Endpoints[j].Addresses[k],
+					TargetRef: items[i].Endpoints[j].TargetRef,
+				}
+				if items[i].Endpoints[j].Hostname != nil {
+					epa.Hostname = *items[i].Endpoints[j].Hostname
+				}
+				addresses = append(addresses, epa)
+			}
+		}
+
+		for j := range items[i].Ports {
+			ports = append(ports, convertEndpointPort(items[i].Ports[j]))
+		}
+
+		subsets = append(subsets, kapi.EndpointSubset{
+			Addresses: addresses,
+			Ports:     ports,
+		})
+	}
```

Comment:
Might be worth preallocating slices, at least `subsets` and `ports`; preallocating `addresses` requires and extra scan, but the extra scan is probably cheaper than reallocs:
```suggestion
	subsets := make([]kapi.EndpointSubset, 0, len(items))
	for i := range items {
		numAddresses := 0
		for j := range items[i].Endpoints {
			numAddresses += len(items[i].Endpoints[j].Addresses)
		}
		addresses := make([]kapi.EndpointAddress, 0, numAddresses)
		n := 0
		for j := range items[i].Endpoints {
			for k := range items[i].Endpoints[j].Addresses {
				epa := kapi.EndpointAddress{
					IP:        ipAddress(items[i].AddressType, items[i].Endpoints[j].Addresses[k]),
					TargetRef: items[i].Endpoints[j].TargetRef,
				}
				if items[i].Endpoints[j].Hostname != nil {
					epa.Hostname = *items[i].Endpoints[j].Hostname
				}
				addresses[n] = epa
				n++
			}
		}

		ports := make([]kapi.EndpointPort, 0, len(items[i].Ports))
		for j := range items[i].Ports {
			ports[j] = convertEndpointPort(items[i].Ports[j])
		}

		subsets[i] = kapi.EndpointSubset{
			Addresses: addresses,
			Ports:     ports,
		}
	}
```

Reply:
> but the extra scan is probably cheaper than reallocs:

I changed my mind on this. Growth is power-of-2 so is it really worth counting all addresses first?